llvm/lib/SYCLLowerIR/SYCLConditionalCallOnDevice.cpp

```diff
@@ -0,0 +1,108 @@
+//===-- SYCLConditionalCallOnDevice.cpp - SYCLConditionalCallOnDevice Pass
+//--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Pass performs transformations on functions which represent the conditional
+// call to application's callable object. The conditional call is based on the
+// SYCL device's aspects or architecture passed to the functions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/SYCLConditionalCallOnDevice.h"
+
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instructions.h"
+
+using namespace llvm;
+
+PreservedAnalyses
+SYCLConditionalCallOnDevicePass::run(Module &M, ModuleAnalysisManager &) {
+  // find call_if_on_device_conditionally function
+  SmallVector<Function *, 4> FCallers;
+  for (Function &F : M.functions()) {
+    if (F.getName().contains("call_if_on_device_conditionally_helper") ||
```

Comment:
Since you are checking for names, you should insert more early exits to reduce amount of string comparisons performed:
- if a function is a kernel (has `spir_kernel` calling convention), we can skip it
- if a function is a declaration, we can skip it